packages/@lwc/template-compiler/src/parser/expression.ts

```diff
@@ -115,33 +115,44 @@ export function parseIdentifier(source: string): TemplateIdentifier | never {
 // Returns the immediate iterator parent if it exists.
 // Traverses up until it finds an element with forOf, or
 // a non-template element without a forOf.
-export function getForOfParent(element: IRElement): IRElement | null {
-    const parent = element.parent;
-    if (!parent) {
-        return null;
-    }
+export function getForOfParent(parentStack: IRElement[]): IRElement | null {
+    let size = parentStack.length;
+    let parent: IRElement | undefined = getParent(parentStack, --size);
+
+    while (parent) {
+        if (parent.forOf) {
+            return parent;
+        }
 
-    if (parent.forOf) {
-        return parent;
-    } else if (parent.tag.toLowerCase() === 'template') {
-        return getForOfParent(parent);
+        parent =
+            parent.tag.toLowerCase() === 'template' ? getParent(parentStack, --size) : undefined;
     }
+
     return null;
 }
 
-export function getForEachParent(element: IRElement): IRElement | null {
-    if (element.forEach) {
-        return element;
-    }
+export function getForEachParent(element: IRElement, parentStack: IRElement[]): IRElement | null {
+    let current: IRElement | undefined = element;
+    let size = parentStack.length;
 
-    const parent = element.parent;
-    if (parent && parent.tag.toLowerCase() === 'template') {
-        return getForEachParent(parent);
+    while (current) {
+        if (current.forEach) {
+            return current;
+        }
+
+        current = getParent(parentStack, --size);
+        if (current?.tag.toLowerCase() !== 'template') {
+            current = undefined;
+        }
     }
 
     return null;
 }
 
-export function isIteratorElement(element: IRElement): boolean {
-    return !!(getForOfParent(element) || getForEachParent(element));
+export function isIteratorElement(element: IRElement, parentStack: IRElement[]): boolean {
+    return !!(getForOfParent(parentStack) || getForEachParent(element, parentStack));
+}
+
+export function getParent(parentStack: IRElement[], size: number): IRElement | undefined {
+    return size > 0 ? parentStack[size] : undefined;
```

Comment:
I don't understand why this is needed. Accessing an array out of bound value should return `undefined`.

Reply:
Great feedback!  I didn't know that about arrays in javascript, TIL!